File: PlagDetection/client/report-viewer/src/model/PlagConfig.ts
interface PlagConfig {
  id?: string;
  name: string;
  subjectName: string;
  resourcesTTL: number;
  clientBackendURL: string;
  projects: StudentProject[];
  scheduleType: ScheduleType;
  userName: string;
}

interface StudentProject {
  name: string;
  resources: Resource[];
  suffixes: Set<string>;
  languages: Set<string>;
  resultHashes: Set<string>;
}

interface Resource {
  exceptionMessage?: string;
  resourceType: ResourceType;
  lastUpdate?: Date;
  path: string;
  academicYear: string;
  hash?: string;
  hasBeenChanged?: boolean;
}

interface GitResource extends Resource {
  repoURL: string;
  branch: string;
}

enum ScheduleType {
  DAILY = 1,
  WEEKLY = 7,
  EVERY_TWO_WEEKS = 14,
  MONTHLY = 30,
}

export enum ResourceType {
  EDGAR = "EDGAR",
  GIT = "GIT",
  ZIP = "ZIP",
}

File: PlagDetection/src/main/kotlin/com/fer/projekt/models/PlagConfig.kt
package com.fer.projekt.models

import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.util.*

@Document(collection = "plag_job_configs")
data class PlagConfig(
    @Id val id: String? = null,
    val name: String,
    val subjectName: String,
    val resourcesTTL: Long,
    val clientBackendURL: String,
    val projects: List<StudentProject>,
    val scheduleType: ScheduleType,
    val userName: String,
    val suffixes: Set<String>,
    val languages: Set<String>,
    val lastUpdateTime: Date? = null,
    val subjectId: String,
    val disallowedFiles: Set<String>,
)

data class StudentProject(
    val name: String,
    val resources: List<Resource>,
    val suffixes: Set<String>,
    val languages: Set<String>,
    val firstRun: Boolean = true,
    val taskText: String? = null,
    val resultHashes: Set<String>?,
)

open class Resource(
    open val exceptionMessage: String? = null,
    open val resourceType: ResourceType,
    open val lastUpdate: Date? = null,
    open val path: String,
    open val academicYear: String,
    open val hash: String? = null,
    open val hasBeenChanged: Boolean? = null,
    val repoURL: String? = null,
    val branch: String? = null,
    val taskText: String? = null,
    val edgarHash: String?
)

enum class ScheduleType(val days: Int) {
    DAILY(1),
    WEEKLY(7),
    EVERY_TWO_WEEKS(14),
    MONTHLY(30),
}

enum class ResourceType {
    EDGAR,
    GIT,
    ZIP,
    WHITELIST,
    AI
}
